crypto_adapters/t_cose_openssl_crypto.c

```diff
@@ -1520,6 +1527,240 @@ aead_byte_count(const int32_t cose_algorithm_id,
+        /* Called in length calculation mode. Return length & exit. */
+        ciphertext->len = expected_output_length;
+        return_value = T_COSE_SUCCESS;
+        goto Done3;
+    }
+
+    /* ------- Algorithm and key and IV length checks -------*/
+    switch(cose_algorithm_id) {
+        case T_COSE_ALGORITHM_A128CTR: evp_cipher = EVP_aes_128_ctr();break;
+        case T_COSE_ALGORITHM_A192CTR: evp_cipher = EVP_aes_192_ctr();break;
+        case T_COSE_ALGORITHM_A256CTR: evp_cipher = EVP_aes_256_ctr();break;
+        case T_COSE_ALGORITHM_A128CBC: evp_cipher = EVP_aes_128_cbc();break;
+        case T_COSE_ALGORITHM_A192CBC: evp_cipher = EVP_aes_192_cbc();break;
+        case T_COSE_ALGORITHM_A256CBC: evp_cipher = EVP_aes_256_cbc();break;
+        default: return_value = T_COSE_ERR_UNSUPPORTED_CIPHER_ALG; goto Done3;
+    }
+    /* This is a sanity check. OpenSSL doesn't provide this check when
+     * using a key. It just assume you've provided the right key
+     * length to EVP_EncryptInit(). A bit unhygenic if you ask me.
+     * Assuming that EVP_CIPHER_key_length() will always return a
+     * small positive integer so the cast to size_t is safe. */
+    expected_key_length = EVP_CIPHER_key_length(evp_cipher);
+    if(key.key.buffer.len != (size_t)expected_key_length) {
+        return_value = T_COSE_ERR_WRONG_TYPE_OF_KEY;
+        goto Done2;
+    }
+    /* Same hygene check for IV/nonce length as for key */
+    /* Assume that EVP_CIPHER_iv_length() won't ever return something
+     * dumb like -1. It would be a bug in OpenSSL or such if it did.
+     * This make the cast to size_t mostly safe. */
+    expected_iv_length = (size_t)EVP_CIPHER_iv_length(evp_cipher);
+    if(nonce.len < expected_iv_length){
+        return_value = T_COSE_ERR_ENCRYPT_FAIL;
+        goto Done2;
+    }
+
+    /* -------- Context initialization with key and IV ---------- */
+    evp_context = EVP_CIPHER_CTX_new();
+    if(evp_context == NULL) {
+        return_value = T_COSE_ERR_ENCRYPT_FAIL;
+        goto Done2;
+    }
+    ossl_result = EVP_EncryptInit_ex(evp_context,
+                                     evp_cipher,
+                                     NULL,
+                                     key.key.buffer.ptr,
+                                     nonce.ptr);
+    if(ossl_result != 1) {
+        return_value = T_COSE_ERR_ENCRYPT_FAIL;
+        goto Done1;
+    }
+
+    /* ---------- Since this is non AEAD cipher AAD is ignored ---------- */
+
+    /* ---------- Actual encryption of plaintext to cipher text ---------*/
+    if(!is_size_t_to_int_cast_ok(expected_output_length)) {
+        /* This tells us that it is not safe to track the output
+         * of the encryption in the integer variables buffer_bytes_used
+         * and bytes_output. */
+        return_value = T_COSE_ERR_ENCRYPT_FAIL;
+        goto Done1;
+    }
+
+    buffer_bytes_used = 0;
+    /* This assumes a stream cipher and no need for handling blocks.
```

Comment:
AES-CBC is not a stream cipher, so the comment is partly wrong.

I would also feel much more comfortable with this if there was a test for for to verify this code works correctly for various sizes that are not on block boundaries. I'm more worried about OpenSSL than PSA.

Reply:
I did more checking and this seems OK. The esdh_enc_dec() test successfully encrypts with a payload of 19 bytes for both PSA and OpenSSL.

PSA clearly uses the PKCS7 padding mode of AES CBC which clearly is for this purpose. I couldn't find documentation for OSSL for this, but it tests correctly in esdh_enc_dec().

So just up the comment to say that it works with streaming and with block modes that have a padding scheme like PKCS7 padding.